apps/glusterfs/volume_entry_test.go

```diff
@@ -2530,3 +2530,67 @@ func TestVolumeCreateMultiClusterErrorsDevices(t *testing.T) {
 			etext)
 	}
 }
+
+func TestVolumeCreateRollbackSpaceReclaimed(t *testing.T) {
+	tmpfile := tests.Tempfile()
+	defer os.Remove(tmpfile)
+
+	// Create the app
+	app := NewTestApp(tmpfile)
+	defer app.Close()
+
+	mockerror := errors.New("MOCK")
+	app.xo.MockVolumeCreate = func(host string, volume *executors.VolumeRequest) (*executors.Volume, error) {
+		return nil, mockerror
+	}
+
+	err := setupSampleDbWithTopology(app,
+		2,    // clusters
+		3,    // nodes_per_cluster
+		4,    // devices_per_node,
+		6*TB, // disksize)
+	)
+	tests.Assert(t, err == nil, "expected err == nil, got", err)
+
+	req := &api.VolumeCreateRequest{}
+	req.Size = 1024
+	req.Durability.Type = api.DurabilityReplicate
+	req.Durability.Replicate.Replica = 3
+
+	vol := NewVolumeEntryFromRequest(req)
+	vc := NewVolumeCreateOperation(vol, app.db)
+
+	e := vc.Build()
+	tests.Assert(t, e == nil, "expected e == nil, got", e)
+
+	e = vc.Exec(app.executor)
+	tests.Assert(t, e != nil, "expected e == nil, got", e)
+
+	e = vc.Rollback(app.executor)
+	tests.Assert(t, e == nil, "expected e != nil, got", e)
+
+	app.db.View(func(tx *bolt.Tx) error {
+		pol, e := PendingOperationList(tx)
+		tests.Assert(t, e == nil, "expected e == nil, got", e)
+		tests.Assert(t, len(pol) == 0, "expected len(pol) == 0, got", len(pol))
+		return nil
+	})
+
+	err = app.db.View(func(tx *bolt.Tx) error {
+		devices, err := DeviceList(tx)
+		if err != nil {
+			return err
+		}
+
+		for _, id := range devices {
+			device, err := NewDeviceEntryFromId(tx, id)
+			if err != nil {
+				return err
+			}
+			tests.Assert(t, device.Info.Storage.Free == 6*TB, id, device)
+		}
+
+		return err
+	})
+	tests.Assert(t, err == nil)
```

Comment:
Generally, the test code asserts that errors were not hit within the callback vs afterwards. Not a big deal though.